### PR TITLE
Line analysis subsets

### PR DESCRIPTION
**Pull request by rosteen**
Separates out the Subset selection from the Data selection in the Line Analysis plugin, in case there are multiple datasets that a subset might be applied to. Also removes some defunct code that was still there, probably from an older version that did things differently.

I'll note here, because it's given us problems before, that I ended up removing the `SubsetCreate` message handling entirely. There is no guarantee of which order message handlers will be executed in in Glue, so I was hitting an error where the handler was getting called (and trying to retrieve the created subset) before the handler that actually adds the subset to the viewer, which is apparently also triggered off the `SubsetCreate` message. As far as I can tell, the `SubsetUpdate` message always gets triggered after a subset is created, and it works to trigger the behavior we need here only off of that message instead.

### CODE REVIEW

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:25]: other plugins have (recently) been changed to be "Entire Spectrum" instead of "None", so probably should match that here (will also need update in `line_analysis.py` line ~74 and ~92).

**Review by kecnry — CHANGES_REQUESTED** [review]
Seems to work well!

I know its a bit of an edge case and possibly out-of-scope for this fix, but deselecting the data in the data menu does clear the data and subset dropdowns, but does not clear the results.  If that's easy to get in here from the same handlers, that might be nice, otherwise we can file a ticket and deal with it later.

Edit: I would say this is mostly UI-related and so does not need a change log entry.

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (rosteen) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:25]: Good call.

**Comment by rosteen** [comment]
Ugh, retrieving the subsets on creation is breaking a couple tests.

**Comment by pllim** [comment]
```
self = Spectral Region, 1 sub-regions:
  (6000.0 Angstrom, 6000.0 Angstrom) 
```

```
mask = array([], dtype=int64)
```

Maybe the event is firing off too early? 🤔

**Comment by rosteen** [comment]
> Maybe the event is firing off too early? 🤔

Hm, maybe. That region is supposed to be 6000-6400...is a `SubsetUpdated` message getting sent on left bound click AND right bound click?? That would be terrible 😵

**Comment by pllim** [comment]
It might be just so, because you do see the subset region update as you move and resize it... 🤔

**Comment by rosteen** [comment]
This might be holiday brain talking, but it seems ok to ignore an error in the subset retrieval here due to oddly constructed subsets that wouldn't happen in real use, rather than crashing everything due to the Line Analysis plugin. I'm willing to be talked into doing more work on it though 😅

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:78]: won't you still get an error raised here (`self._spectral_subsets.keys()`) if entering the except above?

Comment (rosteen) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:78]: You should always have an empty `dict` at least from when the first data was loaded. Unless someone manages to define a subset before any data is loaded in the app I guess...

**Review by pllim — APPROVED** [review]
Can't say I grok everything here but it works for me. ™️

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (duytnguyendtn) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: While we're in this code, here's a small wording/documentation/clarifying change that might be subjective, so I won't block the merge on it. `_run_functions` is super vague to me in a software engineering context (my first impression was that it was intended to be a callback, which is NOT what this was at all!), so I borrowed some language from [the docs](https://jdaviz.readthedocs.io/en/latest/specviz/plugins.html#line-analysis) to inspire, hopefully, some more clear language? 

```suggestion
    def _calc_stats(self, *args, **kwargs):
        """
        Calculate fitting statistics on the initialized models, fixing any
        parameters marked as such by the user, then update the
        displayed parameters with fit values
```

**Review by duytnguyendtn — APPROVED** [review]
One small wording increment to hopefully increase readability? But I won't block on it. Based on what I'm seeing, Kyle's earlier edge case seems resolved on my end (which was a great catch!) and I'm super happy to see us adopting the clearer paradigm established in the Model Fitting plugin.

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (rosteen) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: Good catch - the docstring isn't even accurate, there are no parameters that the user can fix here. Clearly I copied it from the Model Fitting plugin at some point and forgot to update it!

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: Didn't want to block since this code isn't related to your fix, but, you know, while we're in this part of the code, may as well take the opportunity to tidy up 🤷‍♀️

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: Re: tidy up -- Unless the code coverage is good here, I would defer. Tidying untested code may lead to new bugs.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: If it helps, I did run the change in my local environment and it didn't change any behavior I could see

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: Ah, if you already have the patch, it is possible to PR it against this PR branch, if you have not done so already.

**Comment by pllim** [comment]
@rosteen , need a rebase to fix change log conflict. Thanks!

**Comment by rosteen** [comment]
Working on it!

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py
Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:92]: Ok if we're willing to accommodate a PR fix, I can get started on it! @rosteen feel free to bypass this suggestion if there is a rush to get this in, but I'll get started on a PR fix for the above, with some additional clarifying points that I deferred (unless you're already working on it. Not sure if https://github.com/spacetelescope/jdaviz/pull/1012#issuecomment-1000419086 is referring to the rebase or the above?)

Comment (rosteen) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: I already updated it in the commit I just pushed.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: Perfect!

**Comment by rosteen** [comment]
FYI the last commit addressed @duytnguyendtn 's comments about the function name and docstring, as well as properly resetting the selected subset if data in the viewer is deselected and waiting to set results as visible until they are actually set.

**Comment by duytnguyendtn** [comment]
Checks are done. Since I brought up the non-blocking change, I'll give @pllim the merge rights here if you think we're safe to proceed?

**Comment by pllim** [comment]
🎅

**Comment by duytnguyendtn** [comment]
It's a Christmas miracle!

**Pull request merged**
